[Branch: feature/found] Routing with Found

LorbusChris opened the pull request:
Uses [Found](https://github.com/4Catalyzer/found) for react routing.

Implements Client & Server Side Rendering with Found.

See https://github.com/ctrlplusb/react-universally/pull/182#issuecomment-265633372

~~For now, only supports Client Server Rendering with Found.~~

~~I will update this PR as soon as Found's creator @taion adds an SSR API. 
We can then determine, whether or not code-split-component's createRenderContext() will work without additional changes to it.~~

@ctrlplusb

ctrlplusb commented:
This is great @LorbusChris! Thanks.

I've been following `found` myself.  Had an interesting chat with @taion about it on twitter.  Looks like a really awesome and promising project.  I am definitely open to this PR. :)

ctrlplusb commented:
@LorbusChris I see SSR landed. :)

I would hold off doing any coding on this.  A big update is in the pipeline which we should probably wait for before doing any more branch work.

taion commented:
The Found SSR API is pretty trivial anyway (but it handles waiting for data dependencies for you!).

ctrlplusb commented:
@taion which is pretty darn amazing! :)

ctrlplusb commented:
@LorbusChris v10.0.0 is almost ready to ship. :)

LorbusChris commented:
Good news first: This is working in development now. :tada:

Unfortunately in production I'm getting this fatal error in the node console:
`TypeError: Object is not async iterable`
If I'm not mistaken this has to do with this issue with babel-runtime: [babel/#4783:](https://github.com/eact-universallybabel/babel/issues/4783). Maybe we can use babel-polyfill here for the time being?

@ctrlplusb I'm not sure what your plans are in regards to branches using Found (or any other routing lib for that matter), thereby deviating from the base-stack's RRv4. In the mid-run I'm going to be steering this towards relay (with found-relay) and graphql. First, however, I'll connect this to [keystone](https://github.com/keystonejs/keystone), which provides a neat admin UI and API for managing mongoose models and connecting to MongoDB or Redis. If there's any interest in a boilerplate blog/gallery implementation, I'll happily share :)

FYI @taion 
Thanks to you both for the fundamental work beneath!

Everyone, constructive reviews are highly appreciated!

LorbusChris commented on src/client/index.js:
```diff
@@ -25,9 +30,7 @@ function renderApp(TheApp) {
     render(
```
There is a problem with this Promise

taion commented:
Yeah, it is that bug. Interesting. Pulling in the polyfill explicitly would fix things. Let's see.

ctrlplusb commented:
@LorbusChris great work.  In terms of the branch structure I would prefer if you kept the tech stack as purely "found" for now.  We could introduce extended versions of the branch that could introduce additional technologies (i.e. "relay").  :)

I think have a base "found" branch will be great, as it opens itself up for solving data fetching needs for many libraries (e.g. redux).

ctrlplusb commented:
@taion I think the polyfill.io service allows you to specify additional flags for es6 specific features.  I believe `slate` has a similar need.  Perhaps the babel-polyfill isn't a strict requirement.

ctrlplusb commented:
@LorbusChris @taion 

Try the following:

```html
<script src="https://cdn.polyfill.io/v2/polyfill.js?features=es6"></script>
```

taion commented:
It's not an ES6 feature. It's stage 3.

ctrlplusb commented:
Ah oki. 😟

What is the feature? Out of interest.

taion commented:
It's async iterator support – it's how Found manages displaying loading states while fetching data.

ctrlplusb commented:
Of course yes. Thanks.

LorbusChris commented:
@taion Sorry I'm still in the unclear about this. Is this actionable and can be solved by using the polyfill right now?
AFAIK we're already adding the polyfill in generateHTML(). Would this need a static html with the script tag in it?

ctrlplusb commented:
@LorbusChris Thinking about this. You may just need to add the babel async-to-generator plugin to the project.

LorbusChris commented:
@ctrlplusb Thanks, I'll have a look into that! (probably not before the weekend though)

taion commented:
I think babel-polyfill is the way to go pending the runtime fix. It's the symbol polyfill that's broken, not the transpilation.

codepunkt commented:
@LorbusChris This is the first time i've stumbled over `found`. Is it possible to get a short summary of pros/cons compared to react-router v4? Thanks for opening up this possibility! :)

taion commented:
Found has a similar API to React Router v2/v3, except it includes data fetching as well, has better support for displaying loading states while fetching data/code, and uses Redux internally for state management.

LorbusChris commented:
There are actually two issues here, the latter of which is still remaining:

- FIXED: UnhandledPromiseRejectionWarning
The async middleware function has to be wrapped to catch exceptions.

The second one is a bit trickier and has to do with the aforementioned babel-runtime/core-js issue and the ill-transpiled lib code in found's npm release that comes with it.

To investigate this behavior I have drafted [a minimal ssr example](https://github.com/LorbusChris/found-example) - it works fine without babel-polyfill by simply providing regenerator-runtime!

However, I couldn't get Found's SSR to work with our app middleware, with neither babel-node/babel-polyfill nor regenerator-runtime.

Eventually, I tried creating a patched version of Found which does not use transform-runtime in transpilation. In react-universally this gets it done when using babel-polyfill, but it does also not work with regenerator-runtime, in that case again returning `TypeError: Object is not async iterable` @ `babel-runtime/helpers/asyncIterator.js:35:9`. I still haven't figured out under what exact circumstances this breakage occurs and whether or not it is somehow circumventable. Which leaves us with this:
- HOT-FIXED on [this branch](https://github.com/LorbusChris/react-universally/tree/found-patched) with babel-polyfill and patched version of Found.
- BROKEN on this PR, still. Instead I've prepped the code for a fixed version of Found

@ctrlplusb I rebased onto master, so there's lots of commits. Would you mind syncing incubator/routing/found?

@taion FYI

Any ideas how to best approach this thing?

LorbusChris commented:
@ctrlplusb Sure thing, totally get that. Your doing some great community mangement here btw :smile: 
 
I will create a repo for it once I got the feeling it is working alright. For the time being, I'll simply dev on my fork.

Should I close this then?

ctrlplusb commented:
Yeah, we can close this I reckon :)

ctrlplusb commented:
So, we are gonna keep this open. :)

ctrlplusb commented:
Let's keep this PR running until the issues get resolved.

taion commented:
I think a workaround in babel-runtime is possible, anyway. I'm going to try that at some point.

codepunkt commented:
If this also handles data fetching and integrates with redux - how does it work with apollo/graphql? :)

ctrlplusb commented:
Amazing @LorbusChris!!

I can't wait to try it out. Great work!

ctrlplusb commented:
@codepunkt maybe you could use the client-agnostic APIs of Apollo within the `getData` bindings?

taion commented:
I think that's how I'd do the Apollo integration 👍

ctrlplusb commented:
Hey @LorbusChris we can merge this in as soon as we do the next release from `next`.

ctrlplusb commented on src/shared/components/DemoApp/routeConfig.js:
```diff
@@ -0,0 +1,48 @@
+  return <Component {...props} />;
+}
+
+function CodeSplitHome() {
```
👍

It's nice wrapping it up like this.

ctrlplusb commented on tools/webpack/configFactory.js:
```diff
@@ -125,14 +125,17 @@ export default function webpackConfigFactory(buildOptions: BuildOptions) {
       //   import server from './build/server';
       index: removeEmpty([
+        // Required on the server to fix an issue with Found's transpilation
+        ifNode('core-js/library/fn/symbol'),
```
👍

Well found!  Haha, pardon the pun.

taion replied:
yeah... still need to fix this on the babel side, sigh

ctrlplusb commented:
Really great work here @LorbusChris :)

I'm excited to get this in

LorbusChris commented:
@ctrlplusb very glad to hear that! :)

As there's already a merge conflict with the yarn.lock file, I propose the following:

When the next release is out, I'll update the FEATURE.md accordingly and rebase the branch one last time so all changes can be cleanly merged in on top of that release. 

Now, with you having already reviewed the PR, please do tell if you'd prefer a merger over a rebase. The diff should stay the same, though.

LorbusChris commented:
I believe this is ready to merge now :tada:

ctrlplusb commented:
Let's do it! 👍😀

ctrlplusb commented:
💫💥💖

ctrlplusb commented:
We just need to update the feature branches doc on master now. Thanks for this @LorbusChris !

taion commented:
https://github.com/babel/babel/pull/5195 would also fix the babel-runtime issue, BTW.

ctrlplusb commented:
Awesome news @taion 

Thanks for thinking of us! 🙏

taion commented:
This should work now with the new babel-runtime 6.23.0.
